src/Lorentz/Contracts/BaseDAO/FA2.hs

```diff
@@ -0,0 +1,179 @@
+module Lorentz.Contracts.BaseDAO.FA2
+  ( fa2LorentzContract
+  , emptyStorage
+  ) where
+
+import Text.Show (Show (..))
+import qualified Data.Map.Internal as Map
+
+import Lorentz
+import Lorentz.Contracts.Spec.FA2Interface hiding (name, sender)
+import Michelson.Runtime.GState
+import Util.Markdown
+
+type Operators = BigMap (Address, (Address, TokenId)) ()
+
+type Ledger = BigMap (Address, TokenId) Natural
+
+-- | Storage of the FA2 contract
+data Storage = Storage
+  { sLedger      :: Ledger
+  , sOperators   :: Operators
+  , sTokenAddress :: Address
+  }
+  deriving stock (Generic, Show)
+  deriving anyclass (IsoValue, HasAnnotation)
+
+emptyStorage :: Storage
+emptyStorage = Storage
+  { sLedger =  BigMap $ Map.empty
+  , sOperators = BigMap $ Map.empty
+  , sTokenAddress = genesisAddress
+  }
+
+instance TypeHasDoc Storage where
+  typeDocMdDescription =
+    "Contract's storage holding a big_map with all balances and the operators."
+
+contractDoc :: Markdown
+contractDoc = [md|
+  This documentation describes a smart contract which implements FA2 spec
+  (https://gitlab.com/tzip/tzip/-/blob/667f925471728bb8e81fbd30b93a171e401b6dc1/proposals/tzip-12/tzip-12.md)
+  |]
+
+
+fa2LorentzContract :: Contract Parameter Storage
+fa2LorentzContract = defaultContract $ contractName "FA2 smart contract" $ do
+  doc $ DDescription contractDoc
+  unpair
+  entryCase @Parameter (Proxy @PlainEntrypointsKind)
+    ( #cTransfer /-> transfer
+    , #cBalance_of /-> balanceOf
+    , #cToken_metadata_registry /-> tokenMetadataRegistry
+    , #cPermissions_descriptor /-> permissionsDescriptor
+    , #cUpdate_operators /-> updateOperators
+    )
+
+-- | TODO: Implement missing methods
+transfer :: Entrypoint TransferParams Storage
+transfer = undefined
```

Comment:
`undefined` is bad, it causes a warning and makes it impossible to print the contract. You need something like `drop; nil; pair` (not sure it typechecks, but I hope you got the idea).